The documentation of genesis has of two main parts:
The **User Manual** and the **API Reference**.

## User Manual

The manual contains tutorials, code examples and high level explanations of the
functionality of genesis. It is intended for end users who want to use existing scripts and
applications, and for getting started with genesis.

 <!-- *  @subpage intro on how to set up and use genesis. -->
 *  @subpage setup of genesis.
 *  @subpage tutorials with code examples.

Furthermore, there is an information page for the @subpage dev of genesis for those who want
to contribute to the code base.

## API Reference

The documentation of the application programming interface (API) is intended for users who want
to use genesis as a library or toolkit for their own applications.
It provides detailed references for the symbols defined in genesis:

 *  [Namespaces](namespaces.html)
 *  [Classes](annotated.html)
 *  [Source Files](files.html)

As the API Reference is generated from the doc-blocks in the C++ source code, it is mainly
oriented towards the C++ interface. It is however also suitable for the Python interface.

## Support

For user support, please see our [support forum](http://support.genesis-lib.org/). It is
intended for questions on how to do things with genesis.

For bug reports and feature requests, please
[open an issue on our GitHub page](https://github.com/lczech/genesis/issues).

## Acknowledgements

This work is financially supported by the
[Klaus Tschira Foundation](http://www.klaus-tschira-stiftung.de/) and the
[HITS gGmbH - Heidelberg Institute for Theoretical Studies](http://www.h-its.org).

See the webpage of the [Exelixis Lab](http://exelixis-lab.org/) for our other projects.

## References

Please refer to the following articles for more information on phylogenetic placement of short
reads:

> S. Berger, D. Krompass, and A. Stamatakis,
> **Performance, accuracy, and web server for evolutionary placement of short sequence reads
> under maximum likelihood**,
> *Syst. Biol., vol. 60, no. 3, pp. 291–302, 2011*.

> F. A. Matsen, R. B. Kodner, and E. V. Armbrust,
> **pplacer: linear time maximum-likelihood and Bayesian phylogenetic placement of sequences
> onto a fixed reference tree**,
> *BMC Bioinformatics, vol. 11, no. 1, p. 538, 2010*.

> F. A. Matsen, N. G. Hoffman, A. Gallagher, and A. Stamatakis,
> **A format for phylogenetic placements**,
> *PLoS One, vol. 7, no. 2, pp. 1–4, Jan. 2012*.
